### DISCUSSION TITLE
Can we have a way to batch mutate calls?

### DISCUSSION THREAD

**Discussion by ephys** [vercel/swr#472] [General]
Hello,

I'm trying to find a way to prevent SWR from re-rendering on each `mutate` call and instead re-render once after a batch of `mutate` calls

I need this because I'm doing the following:

- create new list item optimistically (mutate)
- scroll document to optimistically created list item
- call API to create the element for real
- rollback the optimistic changes (mutate)
- create the real list item (mutate)

If I don't batch the last two changes, the item will be removed & re-added to the layout, causing the scroll position to shift

I can, instead of rollback then create, update the previously created element. But it makes the code way, way more complicated than just calling my "add item to list" function twice

For reference, this is what the code to add that list item looks like: https://gist.github.com/Ephys/5871cc905c03781bfa6356558ccaedba

**Comment by sergiodxa** [comment]
Instead of removing the optimistic change and adding the new one, you can replace it in a single mutate call or run mutate without data to trigger a revalidation, that will fetch the list again, this time with the new item, and it will replace the current list.

**Reply by ephys** [reply]
Thanks for taking the time to respond @sergiodxa 
Replacing in a single mutate call is something I mentioned but it makes the code much more complex as it's impossible to just reuse the optimistic updater. That is specifically why I asked if adding a way to batch calls can be considered

I also disagree with the concept of calling a second endpoint to refresh when the first API call already returns the data, it's wasteful
Plus in paginated endpoints, we have to make a call for each page that was already loaded (as the chat message is added at the beginning of the list, not the end) which is even more wasteful. Kind of the same issue as with [Partial revalidations](https://github.com/vercel/swr/discussions/465)
